numba/np/ufunc/deviceufunc.py

```diff
@@ -732,110 +724,164 @@ def _broadcast_scalar_input(self, ary, shape):
+        'inputs',
+        '_copy_result_to_host',
     ]
 
+    @abstractmethod
+    def launch_kernel(self, kernel, nelem, args):
+        """Implement the kernel launch"""
+
+    @abstractmethod
+    def is_device_array(self, obj):
+        """
+        Return True if `obj` is a device array for this target
+        """
+
+    @abstractmethod
+    def as_device_array(self, obj):
+        """
+        Return `obj` as a device array on this target.
+
+        May return `obj` directly if it is already on the target.
+        """
+
+    @abstractmethod
+    def to_device(self, hostary):
+        """
+        Copy `hostary` to the device and return the device array.
+        """
+
+    @abstractmethod
+    def device_array(self, shape, dtype):
+        """
+        Allocate a new uninitialized device array with the given shape and
+        dtype.
+        """
+
     def __init__(self, nin, nout, args, kwargs):
-        if nout > 1:
-            raise ValueError('multiple output is not supported')
-        self.args = args
-        self.kwargs = kwargs
-
-        user_output_is_device = False
-        self.output = self.kwargs.get('out')
-        if self.output is not None:
-            user_output_is_device = self.is_device_array(self.output)
-            if user_output_is_device:
-                self.output = self.as_device_array(self.output)
-        self._is_device_array = [self.is_device_array(a) for a in self.args]
-        self._need_device_conversion = (not any(self._is_device_array) and
-                                        not user_output_is_device)
-
-        # Normalize inputs
-        inputs = []
-        for a, isdev in zip(self.args, self._is_device_array):
-            if isdev:
-                inputs.append(self.as_device_array(a))
+        # Get outputs either from the out kwarg, or generate an initial list of
+        # "placeholder" outputs using None as a sentry value
+        outputs = kwargs.get('out')
+        if outputs is not None and len(args) > nin:
+            raise ValueError("cannot specify 'out' as both positional "
+                             "and keyword arguments")
+        else:
+            outputs = [outputs] * nout
+
+        # Ensure all output device arrays are Numba device arrays - any output
+        # passed in that supports the CUDA Array Interface is converted; others
+        # are left untouched.
+        all_user_outputs_are_host = True
+        self.outputs = []
+        for output in outputs:
+            if self.is_device_array(output):
+                self.outputs.append(self.as_device_array(output))
+                all_user_outputs_are_host = False
             else:
-                inputs.append(np.asarray(a))
-        self.norm_inputs = inputs[:nin]
+                self.outputs.append(output)
+
+        all_host_arrays = not any([self.is_device_array(a) for a in args])
+
+        # - If any of the arguments are device arrays, we leave the output on
+        #   the device.
```

Comment:
Is this documented?

Reply:
After some painstaking exploration, I determined that this is a succinct way of expressing what appears to be the behaviour. I think I don't want to document it, in order to leave the door open to figuring out sensible semantics and then documenting them, rather than encoding something that feels evolved then having to deprecate it later.

Reply:
Basically my approach here was "try to make the code tidy / understandable / logical" without breaking any existing tests.

Reply:
Thanks for taking a look at this and the explanation above. I agree with your view on working out what makes sense and then documenting that. That the existing tests pass and the logic is improved should be the target in this PR.